File: problems/cpp/maximum-product-subarray.cc
// 3 ms, 10/09/2017
// Time  : O(n)
// Space : O(n)
// TODO  : Solve it with O(1) space
class Solution {
public:
    int maxProduct(vector<int>& nums)
    {
        if (nums.size() < 2)
            return nums.empty() ? 0 : nums[0];

        /*
           Compute the max subarray product ending at index i using
           the DP relation:
             prod_pos[i] = nums[i] * max(1, prod_pos[i - 1]) if nums[i] > 0 else nums[i] * prod_neg[i - 1]
             prod_neg[i] = nums[i] * max(1, prod_pos[i - 1]) if nums[i] < 0 else nums[i] * prod_neg[i - 1]
           where,
             prod_pos[i] is the max magnitude +ve value prod of subarray with i as the last element, and
             prod_neg[i] is the max magnitude -ve value prod of subarray with i as the last element
        */

        vector<int> prod_pos(nums.size()), prod_neg(nums.size());
        if (nums[0] > 0)
            prod_pos[0] = nums[0], prod_neg[0] = 0;
        else
            prod_neg[0] = nums[0], prod_pos[0] = 0;
        
        int max_prod = prod_pos[0];
        for (int i = 1, n = nums.size(); i < n; ++i) {
            prod_pos[i] = nums[i] > 0 ? nums[i] * max(1, prod_pos[i - 1]) : nums[i] * prod_neg[i - 1];
            prod_neg[i] = nums[i] < 0 ? nums[i] * max(1, prod_pos[i - 1]) : nums[i] * prod_neg[i - 1];
            max_prod = max(max_prod, prod_pos[i]);
        }
        return max_prod;
    }
};

File: code/python/maximum-product-subarray.py
# Date  : 03/18/2018
# Time  : O(n)
# Space : O(n) this implementation; can easily be made O(1)

class Solution:
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        # N[i] stores the min product, -ve if one exists, with A[i] as the last element
        # P[i] stores the max product, +ve if one exists, with A[i] as the last element
        N, P = [1], [1]
        for x in nums:
            p_last, n_last = P[-1], N[-1]
            P.append(max(n_last * x, x) if x < 0 else max(p_last * x, x))
            N.append(min(p_last * x, x) if x < 0 else min(n_last * x, x))

        return max(P[1:])
